### PR TITLE
Allow basepath to be configurable for requests

### PR DESCRIPTION
**Pull request by HenryNguyen5**
- Gets rid of deprecated [api usage](https://nodejs.org/api/url.html#url_legacy_url_api)
- Allows base path for explorer and chainlink clients to be configurable via url
- Always uses absolute URLs during xhr requests

### CODE REVIEW

File: core/cmd/local_client_test.go [hunk 1]
```diff
@@ -28,7 +28,7 @@ func TestClient_RunNodeShowsEnv(t *testing.T) {
 	require.NoError(t, store.KeyStore.Unlock(cltest.Password))
 
 	store.Config.Set("LINK_CONTRACT_ADDRESS", "0x514910771AF9Ca656af840dff83E8264EcF986CA")
-	store.Config.Set("CHAINLINK_PORT", 6688)
+	store.Config.Set("CLIENT_API_PORT", 6688)
```
Comment (HenryNguyen5) [inline core/cmd/local_client_test.go]: This needs to be reverted back.

File: core/cmd/local_client_test.go [hunk 2]
```diff
@@ -69,7 +69,7 @@ func TestClient_RunNodeShowsEnv(t *testing.T) {
 	assert.Contains(t, logs, "LOG_TO_DISK: true")
 	assert.Contains(t, logs, "JSON_CONSOLE: false")
 	assert.Contains(t, logs, "ROOT: /tmp/chainlink_test/")
-	assert.Contains(t, logs, "CHAINLINK_PORT: 6688\\n")
+	assert.Contains(t, logs, "CLIENT_API_PORT: 6688\\n")
```
Comment (HenryNguyen5) [inline core/cmd/local_client_test.go]: This needs to be reverted back.

File: operator_ui/src/api/index.ts [hunk 3]
```diff
@@ -1,4 +1,10 @@
-import * as v2 from './v2'
+import { Api } from '@chainlink/json-api-client'
+import { Sessions } from './sessions'
+import { V2 } from './v2'
 
-export * from './sessions'
-export { v2 }
+const api = new Api({
+  basepath: process.env.CHAINLINK_BASEPATH ?? location.href,
```
Comment (HenryNguyen5) [inline operator_ui/src/api/index.ts]: Should this be set in our docker integration tests?

File: tools/docker/.env [hunk 4]
```diff
@@ -36,6 +36,7 @@ ECHO_SERVER_PORT=6690
 ECHO_SERVER_URL=http://echo-server:6690/count
 
 CHAINLINK_PORT=6688
+CLIENT_API_PORT=6688
```
Comment (HenryNguyen5) [inline tools/docker/.env]: This should be removed. Not used.

File: explorer/client/src/api/v1/index.ts [hunk 5]
```diff
@@ -1,5 +1,14 @@
-import * as adminAuth from './admin/auth'
-import * as adminOperators from './admin/operators'
-import * as jobRuns from './jobRuns'
+import { Api } from '@chainlink/json-api-client'
+import { Auth } from './admin/auth'
+import { Operators } from './admin/operators'
+import { JobRuns } from './jobRuns'
+
+const api = new Api({
+  basepath: process.env.EXPLORER_BASEPATH ?? location.href,
```
Comment (HenryNguyen5) [inline explorer/client/src/api/v1/index.ts]: Should this be set in our docker integration tests?

File: operator_ui/tsconfig.json [hunk 6]
```diff
@@ -9,7 +9,7 @@
     "noEmit": true,
     "allowJs": true
   },
-  "include": ["src", "@types", "test"],
+  "include": ["src", "@types", "__tests__", "support"],
```
Comment (HenryNguyen5) [inline operator_ui/tsconfig.json]: Now we're actually type checking our tests.

File: tools/json-api-client/src/transport/json.ts [hunk 7]
```diff
@@ -65,28 +65,56 @@ export interface PaginatedApiResponse<T extends AttributesObject>
...
+        const options = http.getOptions(method)
+
+        const fetch = fetchWithTimeout(u.toString(), options(params))
+
+        return fetch.then(v => parseResponse(v))
+      }
+    }
+  }
+
+  private createUrl(path: string, query: {} | undefined) {
```
Comment (HenryNguyen5) [inline tools/json-api-client/src/transport/json.ts]: For reviewers: this is the biggest logic change here.

File: operator_ui/src/api/index.ts [hunk 8]
```diff
@@ -1,4 +1,10 @@
-import * as v2 from './v2'
+import { Api } from '@chainlink/json-api-client'
+import { Sessions } from './sessions'
+import { V2 } from './v2'
 
-export * from './sessions'
-export { v2 }
+const api = new Api({
+  base: process.env.CHAINLINK_URL,
```
Comment (rupurt) [inline operator_ui/src/api/index.ts]: How do you feel about renaming this to `CHAINLINK_BASEURL` to keep it consistent?

There are a few integration scripts that would need to be updated.

**Comment by HenryNguyen5** [comment]
> Looking good.
> 
> When building the docker images for core + explorer I feel like it would be best to use the `origin` fallback so that the same docker image can be used for staging & prod. We should only need to explicitly set the `EXPLORER_BASEURL` & `CHAINLINK_BASEURL` when the origin serving the static assets is different to backend.
> 
> How do you feel about adding some explicit documentation to capture that?

Sure, will need to write a couple tests for this.

File: explorer/README.md [hunk 9]
```diff
@@ -55,6 +55,53 @@ $ yarn run dev # in another terminal
 
 Please see [TypeORM's migration guide](https://typeorm.io/#/migrations).
 
+## Running on seperate origins locally
```
Comment (rupurt) [inline explorer/README.md]: 👍

File: explorer/client/src/api/v1/admin/auth.ts [hunk 10]
```diff
@@ -1,32 +1,39 @@
...
+  /**
+   * SignIn authenticates an admin user.
+   * @param username The admin username
+   * @param password The plain text password
+   */
+  @boundMethod
+  public signIn(
+    username: string,
+    password: string,
+  ): Promise<jsonapi.PaginatedApiResponse<number>> {
```
Comment (hellobart) [inline explorer/client/src/api/v1/admin/auth.ts]: Is signIn returns a PaginatedApiResponse type?

Comment (HenryNguyen5) [reply explorer/client/src/api/v1/admin/auth.ts]: It definitely does not, good catch. Mind opening a ticket for this since it's out of scope of this PR?

**Pull request merged**